doc/architectural/memory-primitives.md

```diff
@@ -91,6 +91,69 @@ point into another memory object (such as could happen when running on a real
 machine). To verify that pointers stay within the bounds of their pointees, the
 CBMC option `--pointer-overflow-check` can be used.
 
+### Malloc modelling
+
+CBMC ships a model of `malloc` that seeks to emulate the behaviour of the C
+standard library. In an actual execution, memory allocation may fail for a
+number of reasons and `malloc` would return a NULL pointer. CBMC's model,
+however, will always succeed to allocate an object and never return the `NULL`
+pointer, unless configured as described below. This extends to the case of
+`malloc(0)`, which returns a valid pointer to an object. The size of that object
+is zero, implying that any attempt to read from or write to this object will
+result in an out-of-bounds access. The ensuing undefined behaviour can be
+detected by running CBMC with `--bounds-check`.
+
+The model can be extended to fail allocating memory when the requested
+allocation size is too large, or even non-deterministically fail (for any size).
+Any such failure can either result in calls to `malloc` returning `NULL`, or
+reporting such a call as a failed property.
+The following command line options facilitate the above failure configurations:
+
+|Flag                    |  Check                                            |
+|------------------------|---------------------------------------------------|
+| `--malloc-fail-null`   |  return NULL when emulating an allocation failure |
+| `--malloc-fail-assert` |  report a failed property upon allocation failure |
+| `--malloc-may-fail`    |  non-deterministically fail to allocate           |
+
+Note that the use of `--malloc-may-fail` also requires one of
+`--malloc-fail-null` or `--malloc-fail-assert` to be given. The following code
+example demonstrates the effect of these options:
+
+```C
+int error = 0;
+int *p = malloc(sizeof(int));
+if(p != NULL)
+  *p = 42;
+else
+  error = 1;
+```
+
+Under CBMC's default model of `malloc`, the `else` branch is unreachable.
+When running CBMC with `--malloc-fail-null --malloc-may-fail`, `p` would
+non-deterministically be set to `NULL`, making all branches in the above code
+reachable. Some implementations, however, do not use defensive programming in
+the context of calls to `malloc` and unconditionally dereference the returned
+pointer:
+
+```C
+int *p = malloc(sizeof(int));
+*p = 42;
+```
+
+When running CBMC with `--malloc-fail-null --malloc-may-fail`, `p` would
+non-deterministically be set to `NULL`, and the dereference operation in
+`*p = 42` would amount to a `NULL`-pointer dereference. Such undefined behaviour
+would be reported if, and only if, `--pointer-check` is also used.
+`--malloc-fail-null` is best used with code that uses defensive programming.
```

Comment:
I fail to see the use-case for `--malloc-fail-assert`. Either there is none, or the explanation needs to improve.

Reply:
Quite probably my explanation is not very good, but I'm with you in that we could well do without `--malloc-fail-assert` for `--pointer-check` will catch all relevant cases (the only one we'd miss is a call to `malloc` the result of which is never used).

Reply:
I've removed it and will work on a PR to clean up these options.